src/workflow/ArcanistPatchWorkflow.php

```diff
@@ -1247,11 +1254,33 @@ private function pullBaseTagFromStagingArea($id){
     return self::SUCCESS;
   }
 
+  // fetches hash from any remote repository has configured and also from
+  // staging if repo has one
+  private function pullHashFromAllRemotes($hash) {
+    if (!$hash) {
+      return;
+    }
+    $remotes = array();
+    list($success, $message, $staging, $staging_uri) =
+      $this->validateStagingSetup();
+    if ($success) {
+      $remotes[] = $staging_uri;
+    }
+    $repository_api = $this->getRepositoryAPI();
+    list($err, $stdin, $stderr) = $repository_api->execManualLocal('remote');
+    if (!$err) {
+      $remotes = array_merge($remotes, array_filter(explode(PHP_EOL, $stdin)));
+    }
+    foreach ($remotes as $remote) {
+      $repository_api->execManualLocal('fetch %s %s', $remote, $hash);
+    }
```

Comment:
At this place I could check if we already have hash fetched to avoid fetching any extensive list of repositories...
```
$has_base_revision = $repository_api->hasLocalCommit($hash);
```

Reply:
Yes, you should probably exit early, if you've fetched the hash already.